gobblin-example/src/main/java/gobblin/example/wikipedia/WikipediaExtractor.java

```diff
@@ -124,24 +170,70 @@ public void remove() {
   }
 
   public WikipediaExtractor(WorkUnitState workUnitState) throws IOException {
-    LOG.info("WorkUnitState received: " + workUnitState);
+    //LOG.info("WorkUnitState received: " + workUnitState);
 
     this.workUnitState = workUnitState;
-    this.workUnit = workUnitState.getWorkunit();
     this.rootUrl = readProp(WIKIPEDIA_API_ROOTURL, workUnitState);
     this.schema = readProp(WIKIPEDIA_AVRO_SCHEMA, workUnitState);
-    this.requestedTitles = new LinkedList<>(SPLITTER.splitToList(readProp(SOURCE_PAGE_TITLES, workUnitState)));
-    this.revisionsCnt = Integer.parseInt(readProp(SOURCE_REVISIONS_CNT, workUnitState));
-    this.numRequestedTitles = this.requestedTitles.size();
-
-    if (this.requestedTitles.isEmpty()) {
-      this.recordsOfCurrentTitle = new LinkedList<>();
-    } else {
-      String firstTitle = this.requestedTitles.poll();
-      this.recordsOfCurrentTitle = retrievePageRevisions(firstTitle);
+
+    this.batchSize = 5;
+    this.requestedTitle = workUnitState.getProp(ConfigurationKeys.DATASET_URN_KEY);
+
+    this.baseQuery =
+        ImmutableMap.<String, String>builder().put("format", "json").put("action","query").put("prop","revisions").build();
+
+    try {
+      Queue<JsonElement> lastRevision = retrievePageRevisions(ImmutableMap.<String, String>builder().putAll(this.baseQuery)
+          .put("rvprop","ids").put("titles",this.requestedTitle).put("rvlimit","1").build());
+      this.lastRevisionId = lastRevision.isEmpty() ? -1 : parseRevision(lastRevision.poll());
+    } catch (URISyntaxException use) {
+      throw new IOException(use);
+    }
+
+    long baseRevision = workUnitState.getWorkunit().getLowWatermark(LongWatermark.class, new Gson()).getValue();
```

Comment:
Base revision is never `null`, the source sets it to `-1` if it can't find any data.